### ISSUE TITLE
R infers shape with rigid argument

### ISSUE THREAD

**Issue by lxn2** [dmlc/mxnet#4462]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.
Referencing to StackOverflow question [How bind to names in MXNET mx.io.internal.arrayiter() output?](http://stackoverflow.com/questions/41345033/how-bind-to-names-in-mxnet-mx-io-internal-arrayiter-output)

## Environment info
Operating System:
ubuntu-xenial-16.04-amd64

Compiler:
gcc version 5.4.0 20160609

Package used (Python/R/Scala/Julia):
R

MXNet version:
Or if installed from source:
MXNet commit hash (`git rev-parse HEAD`):
0a0341793f0a5fce887e48a31ec5079f6d2c539c

If you are using R package, please provide

R `sessionInfo()`:
`R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.1 LTS`

`locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C`

`attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     `

`other attached packages:
[1] mxnet_0.7`

`loaded via a namespace (and not attached):
 [1] Rcpp_0.12.8       codetools_0.2-14  visNetwork_1.0.3  digest_0.6.10    
 [5] plyr_1.8.4        jsonlite_1.1      magrittr_1.5      scales_0.4.1     
 [9] stringi_1.1.2     data.table_1.10.0 rstudioapi_0.6    DiagrammeR_0.8.4 
[13] tools_3.2.3       stringr_1.1.0     htmlwidgets_0.8   munsell_0.4.3    
[17] igraph_1.0.1      influenceR_0.1.0  colorspace_1.3-2  htmltools_0.3.5`

## Error Message:
>Please paste the full error message, including stack trace.
[19:28:45] /root/mxnet/dmlc-core/include/dmlc/logging.h:235: [19:28:45]  src/symbol/symbol.cc:155: Symbol.InferShapeKeyword argument name data not  found.
Candidate arguments:
    [0]user
    [1]user1_weight
    [2]item
    [3]item1_weight
    [4]label

>Error in symbol$infer.shape(list(...)) : basic_string::resize
> traceback()
6: stop(list(message = "basic_string::resize", call =   symbol$infer.shape(list(...)), 
       cppstack = NULL))
5: .External(list(name = "CppMethod__invoke_notvoid", address = <pointer: 0x11caef0>, 
       dll = list(name = "Rcpp", path = "/usr/local/lib/R/site-library/Rcpp/libs/Rcpp.so", 
           dynamicLookup = TRUE, handle = <pointer: 0x1404520>, 
           info = <pointer: 0x7f348a598d80>), numParameters = -1L), 
       <pointer: 0x3835820>, <pointer: 0x3836540>, .pointer, ...)
4: symbol$infer.shape(list(...))
3: mx.symbol.infer.shape(symbol, data = input.shape)
2: mx.model.init.params(symbol, input.shape, initializer, mx.cpu())
1: mx.model.FeedForward.create(pred3, X = mx.io.arrayiter(data = DF_mat_x, 
        label = t(data.matrix(data.frame(score = DF_y)))), ctx = devices, 
        num.round = 10, array.batch.size = 10, verbose = T, initializer = mx.init.uniform(0.07), 
       learning.rate = 0.07, eval.metric = mx.metric.rmse, momentum = 0.9, 
       epoch.end.callback = mx.callback.log.train.metric(1))

## Minimum reproducible example
if you are using your own code, please provide a short script that reproduces the error.

>library(mxnet)
DF <- read.delim("./data/ml-100k/u.data", header=F)
names(DF) <- c("user", "item", "score", "time")
max_user<- max(DF$user)
max_item <- max(DF$item)
DF_mat_x <- data.matrix(t(sapply(DF[,1:2], as.numeric)))
DF_y <- DF[,3]
k <- 64
user <- mx.symbol.Variable("user")
item <- mx.symbol.Variable("item")
score <- mx.symbol.Variable("label")
user1 <- mx.symbol.Embedding(data=user, input_dim = max_user, output_dim = k, name="user1") 
item1 <- mx.symbol.Embedding(data=item, input_dim = max_item, output_dim = k, name="item1")
pred <- user1 * item1
pred1 <- mx.symbol.sum_axis(pred, axis = 1, name="pred1")
pred2 <- mx.symbol.Flatten(pred1, name="pred2")
pred3 <- mx.symbol.LinearRegressionOutput(data=pred2, label=score, name="pred3")
devices <- mx.cpu()
mx.set.seed(123)
mx.model.FeedForward.create(pred3, X =  mx.io.arrayiter(data=DF_mat_x,   label=t(data.matrix(data.frame(score=DF_y)))),
                            ctx=devices, num.round=10, array.batch.size=10,
                            verbose=T, #array.layout="rowmajor",
                            initializer=mx.init.uniform(0.07),  learning.rate=0.07,
                            eval.metric=mx.metric.rmse, momentum = 0.9,
                            epoch.end.callback=mx.callback.log.train.metric(1))

## What have you tried to solve it?

1. Changing user <- mx.symbol.Variable("user") to user <- mx.symbol.Variable("data") which returns error
>Error in mx.model.init.params(symbol, input.shape, initializer, mx.cpu()) : Not enough information to get shape

The argument **data=input.shape** seems to be the culprit here. It should be more like [Python's DataIter](https://github.com/dmlc/mxnet-notebooks/blob/master/python/tutorials/matrix_factorization.ipynb) which allows dynamic arguments like "user", "item", and "label" in this case rather than just "data".

```
mx.model.init.params <- function(symbol, input.shape, initializer, ctx) {
  if (!is.MXSymbol(symbol)) stop("symbol need to be MXSymbol")
  slist <- mx.symbol.infer.shape(symbol, data=input.shape)
  if (is.null(slist)) stop("Not enough information to get shapes")
  arg.params <- mx.init.create(initializer, slist$arg.shapes, ctx, skip.unknown=TRUE)
  aux.params <- mx.init.create(initializer, slist$aux.shapes, ctx, skip.unknown=FALSE)
  return(list(arg.params=arg.params, aux.params=aux.params))
}
```

**Comment by piiswrong** [comment]
@thirdwing @hetong007

**Comment by thirdwing** [comment]
@lxn2 You are correct. `data` and `label` shouldn't be hard-coded.

**Comment by thirdwing** [comment]
Fixed in #6673
